File: Solutions/Main/NhsCui.Toolkit.Web/DateInputBoxControl/DateInputBoxValidator.js
//-----------------------------------------------------------------------
// <copyright file="DateInputBoxValidator.js" company="Microsoft Corporation and Crown copyright 2007 - 2010.">
// Copyright (c) Microsoft Corporation and Crown copyright 2007 - 2010.
// All rights reserved.
//
// CERTAIN PARTS OF THIS WORK CONTAIN SOFTWARE CODE THAT IS LICENSED
// FOR USE UNDER THE MICROSOFT PUBLIC LICENSE. DISTRIBUTION, IN SOURCE CODE
// OR OBJECT CODE FORM, OF THOSE PARTS MUST COMPLY WITH THE TERMS OF THE
// PUBLIC LICENSE. SEE http://www.microsoft.com/opensource/licenses.mspx
// FOR DETAILS.
// IF YOU BRING A PATENT CLAIM AGAINST ANY CONTRIBUTOR OVER PATENTS THAT
// YOU CLAIM ARE INFRINGED BY THE PUBLIC LICENSE SOFTWARE, YOUR PATENT
// LICENSE FROM SUCH CONTRIBUTOR TO THE SOFTWARE ENDS AUTOMATICALLY.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY 
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
// </copyright>
// <date>03-Jan-2007</date>
// <summary>Client-side javascript for nhs date inputbox validator</summary>
//-----------------------------------------------------------------------

function DateInputBoxValidatorEvaluateIsValid(val) {
///<summary>Ask NhsCui.DateInputBoxValidator to evaluate validity of the control it is validating
    var value = ValidatorGetValue(val.controltovalidate);
    
    var dateInputBoxComponent = $get(val.controltovalidate).DateInputBox;
    
    //1. Control has validated the value. return true.
    if (dateInputBoxComponent._validatedValue == true) {
        dateInputBoxComponent._validatedValue = false;
        return true;
    }
    
    //2. don't invalidate values like Today, Tomorrow if blur is not occured.
    if (! dateInputBoxComponent._blurOccured && 
    value == dateInputBoxComponent.get_value().toString(dateInputBoxComponent.get_displayDayOfWeek(), false, dateInputBoxComponent.get_displayDateAsText()))
        return true;
    
    //3. If empty return valid 
    if (ValidatorTrim(value).length == 0){
        return true;
    }
    
    //The InputBox element may have a value but it may be the Watermark if so it is "empty" so leave it
    if (val.watermarktext === ValidatorTrim(value)){
        //Essentially box is empty so return valid
        return true;
    }
    
    //4. check if value is null string
    if (dateInputBoxComponent._valueIsNullString === true){
         return true;
    }
    
    //5. Try parsing the date. If parse is successful return true
    try{
        var date = NhsDate.tryParse(value);
                        
        if(date){
        
            if (date.get_dateType() === DateType.Exact || 
                                    date.get_dateType() === DateType.Approximate){
                //6. Check date bounds
            
                if (date.get_dateValue().getTime() < NhsCui.Toolkit.Web.DateInputBox.MinDateValue ||
                                date.get_dateValue().getTime() > NhsCui.Toolkit.Web.DateInputBox.MaxDateValue){
                    return false;
                }
                else{
                    return true;
                }
            }
            else if (date.get_dateType() === DateType.YearMonth || 
                                    date.get_dateType() === DateType.Year){
                 //7. Check year bounds
            
                if (date.get_year() < new Date(NhsCui.Toolkit.Web.DateInputBox.MinDateValue).getFullYear() ||
                                date.get_year() > new Date(NhsCui.Toolkit.Web.DateInputBox.MaxDateValue).getFullYear()){
                    return false;
                }
                else{
                    return true;
                }                   
            }
            else{
                return true;
            }
        }
        else{
            return false;
        }
    }
    catch(e){
        return false;
    }
}

File: Solutions/Main/NhsCui.Toolkit.Web/DateInputBoxControl/DateInputBoxValidator.cs
//-----------------------------------------------------------------------
// <copyright file="DateInputBoxValidator.cs" company="Microsoft Corporation and Crown copyright 2007 - 2010.">
// Copyright (c) Microsoft Corporation and Crown copyright 2007 - 2010.
// All rights reserved.
//
// CERTAIN PARTS OF THIS WORK CONTAIN SOFTWARE CODE THAT IS LICENSED 
// FOR USE UNDER THE MICROSOFT PUBLIC LICENSE. DISTRIBUTION, IN SOURCE CODE 
// OR OBJECT CODE FORM, OF THOSE PARTS MUST COMPLY WITH THE TERMS OF THE 
// PUBLIC LICENSE. SEE http://www.microsoft.com/opensource/licenses.mspx 
// FOR DETAILS.  
// IF YOU BRING A PATENT CLAIM AGAINST ANY CONTRIBUTOR OVER PATENTS THAT 
// YOU CLAIM ARE INFRINGED BY THE PUBLIC LICENSE SOFTWARE, YOUR PATENT 
// LICENSE FROM SUCH CONTRIBUTOR TO THE SOFTWARE ENDS AUTOMATICALLY.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY 
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
// </copyright>
// <date>16-May-2007</date>
// <summary>Validation control for the NhsCui.Toolkit.Web.DateInputBox</summary>
//-----------------------------------------------------------------------
[assembly: System.Web.UI.WebResource("NhsCui.Toolkit.Web.DateInputBoxControl.DateInputBoxValidator.js", "application/x-javascript")]
namespace NhsCui.Toolkit.Web
{
    using System;
    using System.Drawing;
    using System.Collections.Generic;
    using System.Text;
    using System.Web;
    using System.Web.UI;
    using System.Web.UI.WebControls;
    using System.ComponentModel;
    using NhsCui.Toolkit.DateAndTime;
    using NhsCui.Toolkit.Web;
    using System.Globalization;
            
    /// <summary>
    /// Validation control for the NhsCui.Toolkit.Web.DateInputBox
    /// </summary>
    [ToolboxItemFilterAttribute("System.Web.UI", ToolboxItemFilterType.Require)]
    [ToolboxBitmap(typeof(ToolboxBitmaps.ResourceReference), "DateInputBoxValidator.bmp")]
    [ToolboxData("<{0}:DateInputBoxValidator runat=\"server\" ErrorMessage=\"DateInputBoxValidator\"></{0}:DateInputBoxValidator>")]
    public class DateInputBoxValidator : BaseValidator
    {
        /// <summary>
        /// Holds the watermark text
        /// </summary>
        private string watermarkText = "dd-MMM-yyyy";
        
        /// <summary>
        /// The DateInputBox uses a Watermark Extender which places text into the DateInputBox when it is empty.
        /// This text is more than likely an Invalid date 
        /// </summary>
        [Category("Behavior")]
        [Description("The DateInputBox uses a Watermark Extender which places text into the DateInputBox when it is empty. Use this Text as aprt of the IsEmpty check")]
        [DefaultValue("dd-MMM-yyyy")]
        public string WatermarkText
        {
            get
            {
                return this.watermarkText;
            }

            set
            {
                this.watermarkText = value;
            }
        }
        
        /// <summary>
        /// Determines whether the control specified by the ControlToValidate property is a valid control.
        /// </summary>
        /// <returns>True if control properties is valid</returns>
        protected override bool ControlPropertiesValid()
        {
            // Call the base implementation of ControlPropertiesValid first. 
            // If that passes run the extra checks 
            
            if (base.ControlPropertiesValid() == true)
            {
                DateInputBox control = this.NamingContainer.FindControl(this.ControlToValidate) as DateInputBox;

                if (control == null)
                {
                    throw new HttpException(DateInputBoxControl.DateInputBoxValidatorResources.ControlToValidateInvalid);
                }
                else
                {
                    return true;
                }
            }
            else
            {
                return false;
            }
        }

        /// <summary>
        /// Check if the ControlToValidate is valid
        /// </summary>
        /// <returns>True if control properties is valid</returns>
        protected override bool EvaluateIsValid()
        {
            string value = this.GetControlValidationValue(this.ControlToValidate);            
            DateInputBox control = this.NamingContainer.FindControl(this.ControlToValidate) as DateInputBox;
            if (value.Length == 0)
            {
                return true;
            }

            if (value == this.WatermarkText)
            {
                // Essentially box is empty so return valid
                return true;
            }

            // Don't raise an error for Today, Tomorrow if value is also the same.
            if (value == control.Value.ToString(control.DisplayDayOfWeek, false, control.DisplayDateAsText, CultureInfo.CurrentCulture))
            {
                return true;
            }

            NhsDate tryParseResult;

            if (!NhsDate.TryParseExact(value, out tryParseResult, System.Threading.Thread.CurrentThread.CurrentCulture))
            {
                control.Value.DateType = DateType.Null;
                return false;                                
            }

            return true;
        }

        /// <summary>
        /// PreRender time processing
        /// </summary>
        /// <param name="e">e</param>
        protected override void OnPreRender(EventArgs e)
        {
            if (this.EnableClientScript == true)
            {
                if (HttpContext.Current.Request.Browser.EcmaScriptVersion.Major >= 1)
                {
                    this.RegisterClientScript();
                }
            }
            
            base.OnPreRender(e);
        }

        /// <summary>
        /// Set up client script 
        /// </summary>
        private void RegisterClientScript()
        {
            this.Page.ClientScript.RegisterExpandoAttribute(this.ClientID, "evaluationfunction", "DateInputBoxValidatorEvaluateIsValid");
            this.Page.ClientScript.RegisterExpandoAttribute(this.ClientID, "valtype", "NhsCui.DateInputBoxValidator");
            this.Page.ClientScript.RegisterExpandoAttribute(this.ClientID, "watermarktext", this.WatermarkText);
            
            this.Page.ClientScript.RegisterClientScriptResource(typeof(NhsCui.Toolkit.Web.DateInputBoxValidator), "NhsCui.Toolkit.Web.DateInputBoxControl.DateInputBoxValidator.js");
        }       
    }
}
